pkg/adapter/common/backoff_test.go

```diff
@@ -127,47 +128,89 @@ func TestRun(t *testing.T) {
 			waitReturn: true,
 			err:        assert.AnError,
 		},
-		"fn resets backoff duration": {
-			fn: func(ctx context.Context) (bool, error) {
-				counter++
-				if counter > 3 {
-					if time.Since(start).Round(time.Second) != 3*time.Second {
-						return true, fmt.Errorf("did we reset the backoff?")
-					}
-					// trigger Run() termination with expected error
-					return true, assert.AnError
-				}
-				// reset backoff duration to 1 sec for each iteration
-				return true, nil
-			},
-			waitReturn: true,
-			err:        assert.AnError,
-		},
 	}
 
-	errCh := make(chan error)
-	defer close(errCh)
-
 	for name, tc := range testCases {
-		stopCh := make(chan struct{})
 		t.Run(name, func(t *testing.T) {
-			bo := NewBackoff()
-			go func() {
-				err := bo.Run(stopCh, tc.fn)
-				errCh <- err
-			}()
-			if tc.waitReturn {
-				defer close(stopCh)
-			} else {
-				close(stopCh)
-			}
+			errCh := make(chan error)
+			defer close(errCh)
+
+			stopCh := make(chan struct{})
+
+			t.Run(name, func(t *testing.T) {
+				bo := NewBackoff()
+				go func() {
+					err := bo.Run(stopCh, tc.fn)
+					errCh <- err
+				}()
+
+				if tc.waitReturn {
+					defer close(stopCh)
+				} else {
+					close(stopCh)
+				}
 
-			err := <-errCh
-			if tc.err != nil {
-				assert.EqualError(t, err, tc.err.Error())
-			} else {
-				assert.NoError(t, err)
-			}
+				err := <-errCh
+				if tc.err != nil {
+					assert.EqualError(t, err, tc.err.Error())
+				} else {
+					assert.NoError(t, err)
+				}
+			})
 		})
 	}
+
+	t.Run("fn fails several times then resets backoff duration", func(t *testing.T) {
+		const minBackoff = 1 * time.Millisecond
+
+		const timesRunFnBeforeReset = 3
+		var runCounter int
+
+		var returnVal bool
+
+		// latch is used to keep the function executions under control
+		// of the main test logic.
+		latch := make(chan struct{})
+
+		fn := func(ctx context.Context) (bool, error) {
+			latch <- struct{}{}
+
+			defer func() {
+				runCounter++
+				latch <- struct{}{}
+			}()
```

Comment:
I'm not sure that I completely understand why do we need that deferred function with the counter, but the thing with double locking latches is something that I want to use somewhere else :slightly_smiling_face:

Reply:
Nice catch, this is a leftover from an older version of this code! We only need to defer the "unlatch" here.

Reply:
Removed